### PR TITLE
Enable omitting parts of the documentation toctree when rendering PDF

### PR DESCRIPTION
**Pull request by MattF-NSIDC**
## Description

The Sphinx `only::` directive [isn't capable of modifying the toctree](https://github.com/sphinx-doc/sphinx/issues/2150). 

~This PR introduces Jinja2 templating to enable a different toctree for PDF output (omit contributor docs) than for HTML output (full docs). I put all the logic in our invoke tasks.~ Ignore this, the `jinja2` approach doesn't have a solution for providing the correct index file to the readthedocs build process.

This PR introduces our own extension for filtering content out of toctrees based on the selected builder, and uses a third-party extension for changing the behavior of sphinx's `only::` directive to occur during parse time, making it effective at filtering out whole toctrees based on builder.

This is trying to achieve the same thing as #503, but we had some issues with that approach and wanted to try  templating or something else. This is working well. 

NOTE: When rendering PDF, it's expected that warnings will be produced, so PDF-rendering commands don't treat warnings as errors by overriding `SPHINXOPTS` to an empty string. If the HTML renders with no warnings, we can ignore any PDF warnings.

## Checklist

If an item on this list is done _or not needed_, simply check it with `[x]`.

- [x] Config lockfile updated (`inv config.export > qgreenland/config/cfg-lock.json`)
- [x] Version bumped if needed (`bumpversion (major|minor|patch|prerelease|build`)
- [x] CHANGELOG.md updated
- [x] Documentation updated if needed
- [x] New unit tests if needed

### CODE REVIEW

**Comment by MattF-NSIDC** [comment]
I feel silly for forgetting that we need an `index.rst` for readthedocs. Hm...

**Comment by MattF-NSIDC** [comment]
Maybe instead of using Jinja2 we resort to some simple search-and-replace to omit lines between two comment markers when rendering PDF. Maybe we can write a sphinx extension that can pre-process the file and omit content using a new custom directive?

File: doc/Makefile [hunk 1]
```diff
@@ -4,8 +4,6 @@
 # You can set these variables from the command line, and also
 # from the environment for the first two.
 
-# TODO: Support nitpicky mode in readthedocs
```
Comment (MattF-NSIDC) [inline doc/Makefile:7]: Moved this comment in to `conf.py`; pretty sure we can enable nitpicky mode there.

**Comment by MattF-NSIDC** [comment]
Resolves #524

File: doc/sphinx-ext/toctree_select_builder.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+            prefix = m.groups()[0]
+            value = m.groups()[1]
+
+            # If there's a prefix, keep only if it matches builder name
+            if prefix == builder_name:
+                # Keep the value but not the prefix
+                selected.append(value)
+
+            else:
```
Comment (trey-stafford) [inline doc/sphinx-ext/toctree_select_builder.py:57]: Would it be worth checking that the prefix is a known/expected builder (e.g., `html` or `pdf`)? This way we avoid missing errors like `hmtl:thing_we_really_want_in_html` (`hmtl` should be `html`)

**Review by trey-stafford — APPROVED** [review]
One idea for an additional check in `toctree_select_builder.py` but otherwise looks good! Glad this works!

File: doc/sphinx-ext/toctree_select_builder.py (hunk 2 above)
Comment (MattF-NSIDC) [reply doc/sphinx-ext/toctree_select_builder.py:57]: I like this idea :)

Comment (MattF-NSIDC) [reply doc/sphinx-ext/toctree_select_builder.py:66]: efafa080

**Pull request merged**